libcontainer/cgroups/ebpf/ebpf.go

```diff
@@ -1,29 +1,93 @@
 package ebpf
 
 import (
+	"runtime"
+	"unsafe"
+
 	"github.com/cilium/ebpf"
 	"github.com/cilium/ebpf/asm"
 	"github.com/cilium/ebpf/link"
 	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
 	"golang.org/x/sys/unix"
 )
 
+func nilCloser() error {
+	return nil
+}
+
+func findAttachedCgroupDeviceFilters(dirFd int) ([]*ebpf.Program, error) {
+	type bpfAttrQuery struct {
+		TargetFd    uint32
+		AttachType  uint32
+		QueryType   uint32
+		AttachFlags uint32
+		ProgIds     uint64 // __aligned_u64
+		ProgCnt     uint32
+	}
+
+	// Currently you can only have 64 eBPF programs attached to a cgroup.
+	size := 64
+	allowRetry := true
+	for allowRetry {
+		progIds := make([]uint32, size)
+		query := bpfAttrQuery{
+			TargetFd:   uint32(dirFd),
+			AttachType: uint32(unix.BPF_CGROUP_DEVICE),
+			ProgIds:    uint64(uintptr(unsafe.Pointer(&progIds[0]))),
+			ProgCnt:    uint32(len(progIds)),
+		}
+
+		// Fetch the list of program ids.
+		_, _, errno := unix.Syscall(unix.SYS_BPF,
+			uintptr(unix.BPF_PROG_QUERY),
+			uintptr(unsafe.Pointer(&query)),
+			unsafe.Sizeof(query))
+		size = int(query.ProgCnt)
+		runtime.KeepAlive(query)
+		if errno != 0 {
+			// On ENOSPC we get the correct number of programs.
+			if errno == unix.ENOSPC {
+				allowRetry = false
+				continue
+			}
+			return nil, errno
+		}
+
+		// Convert the ids to program handles.
+		programs := make([]*ebpf.Program, len(progIds))
+		for idx, progId := range progIds {
+			program, err := ebpf.NewProgramFromID(ebpf.ProgramID(progId))
+			if err != nil {
+				return nil, errors.Wrap(err, "cannot fetch program from id")
```

Comment:
The new golang style to wrap an error is `fmt.Errorf("cannot fetch program from id: %w", err)`, but we still have a lot of `errors.Wrap()` so whatever.

Reply:
I'll switch them all to `Errorf("%w")`.

Reply:
What is the problem with `errors.Wrap`?

Reply:
There's nothing wrong with `errors.Wrap` specifically, but the newer `fmt.Errorf("%w")` setup is in the stdlib and the stdlib `errors.As` and `errors.Is` are probably going to end up more widely used (`pkg/errors` does still have the upside that you get a stack trace though...).